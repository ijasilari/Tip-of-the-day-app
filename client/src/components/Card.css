.card-container {
  margin: auto;
  justify-content: center;
  align-items: center;
  height: fit-content;
}

.card {
  background-color: #ffffff;
  color: #333333;
  border-radius: 10px;
  box-shadow: 0px 2px 10px rgba(0, 0, 0, 0.2);
  width: fit-content;
  height: fit-content;
  max-width: 90%;
  display: flex;
  flex-direction: column;
  justify-content: left;
  align-items: left;
  margin: auto;
  padding: 10px 20px;
  overflow: auto;
}

.title {
  margin: auto;
}

.loading {
  font-size: 24px;
  color: #333333;
}

#light .card {
  background-color: #fff;
}

#dark .card {
  background-color: #1c1c1c;
}
