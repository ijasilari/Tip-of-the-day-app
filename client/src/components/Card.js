import React, { useState, useEffect } from "react";
import axios from "axios";
import "./Card.css";
import ReactMarkdown from "react-markdown";
import { Prism as SyntaxHighlighter } from "react-syntax-highlighter";
import { tomorrow } from "react-syntax-highlighter/dist/esm/styles/prism";
import { useAutoAnimate } from "@formkit/auto-animate/react";

const Card = (props) => {
  const [cardData, setCardData] = useState(null);
  const [cardCategory, setCardCategory] = useState("");
  const [dataLoaded, setDataLoaded] = useState(false);

  let textColor = "";
  if (props.theme === "light") {
    textColor = "black";
  } else {
    textColor = "#ECECEC";
  }

  const categoryOptions = [
    { value: 0, label: "Other" },
    { value: 1, label: "CSS" },
    { value: 2, label: "Java" },
    { value: 3, label: "JavaScript" },
    { value: 4, label: "HTTP" },
    { value: 5, label: "Python" },
    { value: 6, label: "CPP" },
    { value: 7, label: "Dart" },
    { value: 8, label: "Flutter" },
    { value: 9, label: "Rust" },
    { value: 10, label: "Linux" },
  ];

  let responseCategoryNum = 0;
  let categoryLabel = "";
  const fetchCardData = async () => {
    const response = await axios.get(
      `${process.env.REACT_APP_LOCAL_BACKEND_URL}/api/tips/randomtip`
    );
    setCardData(response.data.tip);
    setCardCategory(response.data.tip.category);

    responseCategoryNum = response.data.tip.category;
    categoryLabel = categoryOptions[responseCategoryNum];

    setCardCategory(categoryLabel.label);
  };

  useEffect(() => {
    fetchCardData();
    setDataLoaded(true);
  }, []);

  useEffect(() => {
    if (dataLoaded) {
      const interval = setInterval(() => {
        fetchCardData();
      }, 12000);
      return () => clearInterval(interval);
    }
  }, [dataLoaded]);

  const [animationParent] = useAutoAnimate();

  return (
    <div className="card-container">
      {cardData ? (
        <div className="card" ref={animationParent}>
          <h2 className="title">{cardCategory}</h2>
          <ReactMarkdown
            children={cardData.description}
            components={{
              p: ({ node, ...props }) => (
                <p style={{ color: textColor }} {...props} />
              ),
              code({ node, inline, className, children, ...props }) {
                const match = /language-(\w+)/.exec(className || "");
                return !inline && match ? (
                  <SyntaxHighlighter
                    children={String(children).replace(/\n$/, "")}
                    language={match[1]}
                    style={tomorrow}
                    {...props}
                  />
                ) : (
                  <code className={className} {...props}>
                    {children}
                  </code>
                );
              },
            }}
          />
        </div>
      ) : (
        <div className="loading" style={{ textAlign: "center" }}>
          Loading...
        </div>
      )}
    </div>
  );
};

export default Card;
