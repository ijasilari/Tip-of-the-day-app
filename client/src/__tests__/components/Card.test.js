import { render, waitFor, screen } from '@testing-library/react';
import renderer from "react-test-renderer";
import axiosMock from "axios";
import Card from '../../components/Card';

jest.mock("axios");
jest.useFakeTimers();

describe("Card", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it("should render fetched data", async () => {
    const testData = {
      tip: {
        category: 1,
        description: "Some test data",
      },
    };

    axiosMock.get.mockResolvedValueOnce({ data: testData });

    let component;

    await renderer.act(async () => {
      component = renderer.create(<Card />);
    });
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });
  /*it("fetches new data after 12 seconds", async () => {
    const responseData = {
      data:{
        tip:{
          category: 3,
          description: "This is a new tip.",
        }
      }
    };
    axiosMock.get.mockResolvedValueOnce({data: responseData.data.tip.category });

    const { getByText } = render(<Card />);

    expect(screen.getByText("Loading...")).toBeInTheDocument();
    await waitFor(() => expect(screen.queryByText("Loading...")).not.toBeInTheDocument());

    expect(screen.getByText("JavaScript")).toBeInTheDocument();
    expect(screen.getByText("This is a new tip.")).toBeInTheDocument();

    const newResponseData = {
      data:{
        tip: {
          category: 5,
          description: "This is another new tip.",
        }
      }
    };
    axiosMock.get.mockResolvedValueOnce({data: newResponseData.data.tip.category });

    jest.advanceTimersByTime(12000);

    expect(screen.getByText("Loading...")).toBeInTheDocument();
    await waitFor(() => expect(screen.queryByText("Loading...")).not.toBeInTheDocument());

    expect(screen.getByText("Python")).toBeInTheDocument();
    expect(screen.getByText("This is another tip.")).toBeInTheDocument();
  });*/
});


