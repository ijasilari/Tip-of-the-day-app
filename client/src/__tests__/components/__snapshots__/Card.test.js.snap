// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should render fetched data 1`] = `
<div
  className="card-container"
>
  <div
    className="card"
  >
    <h2
      className="title"
    >
      CSS
    </h2>
    <p>
      Some test data
    </p>
  </div>
</div>
`;
